src/main/java/mate.academy/lib/Injector.java

```diff
@@ -1,13 +1,81 @@
 package mate.academy.lib;
 
+import java.lang.reflect.Constructor;
+import java.lang.reflect.Field;
+import java.util.HashMap;
+import java.util.Map;
+import mate.academy.service.FileReaderService;
+import mate.academy.service.ProductParser;
+import mate.academy.service.ProductService;
+import mate.academy.service.impl.FileReaderServiceImpl;
+import mate.academy.service.impl.ProductParserImpl;
+import mate.academy.service.impl.ProductServiceImpl;
+
 public class Injector {
     private static final Injector injector = new Injector();
+    private final Map<Class<?>, Object> instances = new HashMap<>();
+    private final Map<Class<?>, Class<?>> interfaceImplementations = Map.of(
+            ProductService.class, ProductServiceImpl.class,
+            FileReaderService.class, FileReaderServiceImpl.class,
+            ProductParser.class, ProductParserImpl.class
+    );
 
     public static Injector getInjector() {
         return injector;
     }
 
-    public Object getInstance(Class<?> interfaceClazz) {
-        return null;
+    public Object getInstance(Class<?> interfaceClass) {
+        try {
+            if (instances.containsKey(interfaceClass)) {
+                return instances.get(interfaceClass);
+            }
+
+            Class<?> implementationClass = interfaceImplementations.get(interfaceClass);
+            if (implementationClass == null
+                    || !implementationClass.isAnnotationPresent(Component.class)) {
+                throw new IllegalArgumentException(
+                        "Injection failed, missing @Component annotation on the class "
+                                + (implementationClass != null
+                                ? implementationClass.getName() : "null")
+                );
+            }
+
+            Object instance = createInstance(implementationClass);
+            instances.put(interfaceClass, instance);
+            injectDependencies(instance);
+            return instance;
+        } catch (ReflectiveOperationException e) {
+            throw new RuntimeException(
+                    "Injection failed for " + interfaceClass.getName(), e
+            );
+        }
+    }
+
+    private Object createInstance(Class<?> clazz) throws ReflectiveOperationException {
+        Constructor<?>[] constructors = clazz.getDeclaredConstructors();
+        for (Constructor<?> constructor : constructors) {
+            if (constructor.isAnnotationPresent(Inject.class)) {
+                Class<?>[] parameterTypes = constructor.getParameterTypes();
+                Object[] parameters = new Object[parameterTypes.length];
+                for (int i = 0; i < parameterTypes.length; i++) {
+                    parameters[i] = getInstance(parameterTypes[i]);
+                }
+                return constructor.newInstance(parameters);
+            }
+        }
+        return clazz.getDeclaredConstructor().newInstance();
```

Comment:
The createInstance method is using the @Inject annotation to find the appropriate constructor, which aligns with the checklist requirement to use only @Component and @Inject annotations.